extensions/sql-database-projects/src/test/projectController.test.ts

```diff
@@ -98,49 +102,45 @@ describe('ProjectsController: project controller operations', function (): void
 
 		it('Should return silently when no SQL object name provided in prompts', async function (): Promise<void> {
 			for (const name of ['', '    ', undefined]) {
-				testContext.apiWrapper.reset();
-				testContext.apiWrapper.setup(x => x.showInputBox(TypeMoq.It.isAny(), TypeMoq.It.isAny())).returns(() => Promise.resolve(name));
-				testContext.apiWrapper.setup(x => x.showErrorMessage(TypeMoq.It.isAny())).returns((s) => { throw new Error(s); });
```

Comment:
So these calls ended up being where I had to spend the most time debugging and fixing them up. By throwing here you're not only shortcutting any potential logic that might run after the dialog is shown but you're also testing behavior that isn't normal (I wouldn't expect this call to ever actually throw an exception unless things are really broken).

In general let's try to make sure that what we're testing is actually the behavior that's going to be happening at runtime.

Reply:
But by making showErrorMessage throw you're having the code go through a path that we should realistically never hit in most scenarios. 

When you say "short-circuiting" are you referring to the code just returning early if an error state was reached that you were showing a warning/error for? That shouldn't impact the test - you should still have it go through the entire path like it normally would and then at the end ensure that the calls you expected were made.

Looking at this particular test I think I just choose a bad example. All the other cases of this kind of code was checking that showErrorMessage was called with a specific message. For those cases you should just spy and then use calledWith to verify the text.

In this case yes, the correct way in my eyes would just be to check that showErrorMessage was never called. You still want it to go through the rest of the logic (which I assume is properly short-circuiting to return early) to make sure that nothing is breaking in that flow. 

I've updated the test to check for this.

Reply:
> When you say "short-circuiting" are you referring to the code just returning early if an error state was reached that you were showing a warning/error for?

Yup, that's the goal for this particular test.  If there's an exception or `showErrorMessage()` is called for those specific inputs, then the code being tested is broken (because it should be terminating early and silently), and the failure message for the test gets set to the error-that-shouldn't-be.  There's no reason to have the test continue through the rest of the path because it's now in an invalid/broken state.  For most of the rest, I think you're right about spy + calledWith.

Reply:
Yeah but you still want the test to verify that something else doesn't break while it's winding back up the stack. Throwing an exception there means you skip that logic which means you could be missing hitting some other error that happens along the way.